### PR TITLE
Add tests to ref return error codes

### LINKED ISSUE #13073: Add missing unit tests for C# ref return error codes
Add unit tests for the following:

- [x] ERR_BadIteratorReturnRef = 8154,
- [x] ERR_BadRefReturnExpressionTree = 8155,
- [x] ERR_RefReturningCallInExpressionTree = 8156,
- [x] ERR_RefReturnRangeVariable = 8160,
- [x] ERR_RefAssignmentMustHaveIdentityConversion = 8188,
- [x] ERR_BadIteratorLocalType = 8191,
- [x] ERR_BadAsyncLocalType = 8192,
- [x] ERR_RefReturningCallAndAwait = 8193,

### PR DESCRIPTION
**Pull request by OmarTawfik**
Closes #13073

### CODE REVIEW

**Comment by jaredpar** [comment]
@TyOverby @ellismg 

Another case where the taskkill for MSBuild hit us:

``` txt
15:08:11 Killing all build-related processes
15:08:11 ERROR: taskkill for msbuild.exe failed
```

https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_unit64_prtest/2220/console

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 1]
```diff
@@ -1101,5 +1101,256 @@ static void Main(string[] args)
...
+        public void CannotUseYieldReturnInAReturnByRefFunction()
+        {
+            var code = @"
+class TestClass
+{
+    int x = 0;
+    ref int TestFunction()
+    {
+        yield return x;
+    }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Can we add a local function version of this test as well?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 2]
```diff
@@ -1101,5 +1101,256 @@ static void Main(string[] args)
...
+            var code = @"
+using System.Collections.Generic;
+class TestClass
+{
+    int x = 0;
+    IEnumerator<int> TestMethod()
+    {
+        ref int y = ref x;
+        yield return 0;
+    }
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Can we add a  local function version of the test as well?

**Comment by OmarTawfik** [comment]
@jaredpar comments addressed.
@dotnet/roslyn-compiler can I get another review?

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler can I get another review?

**Comment by jcouv** [comment]
On it. Sorry for the delay.

**Pull request merged**